docs/design/CONFIG_API.md

```diff
@@ -0,0 +1,228 @@
+Configuration API Design
+===================
+
+## Motivation
+[Issue 51][issue_51] specifies the need to design an API for configuring Envoy Gateway at runtime. This configuration is
+referred to as the "static configuration" in the Envoy Gateway [design doc][design_doc].
+
+## Goals
+* Define an __initial__ API for configuring Envoy Gateway at runtime, independent of the runtime platform, e.g.
+  Kubernetes.
+
+## Non-Goals
+* Implementation of the Envoy Gateway runtime configuration API.
+* Define the `status` subresource of the runtime configuration API.
+* Define a __complete__ API for configuring Envoy Gateway at runtime. As stated in the [Goals](#goals), this document
+  defines the initial runtime configuration API for Envoy Gateway.
+* Define an API for deploying/provisioning/operating Envoy Gateway.
+* Define an __initial__ provider-independent API for configuring the managed Envoy proxy infrastructure.
+* Specify tooling for managing the API, e.g. generate protos, CRDs, controller RBAC, etc.
+
+## API Definition
+`ControlPlaneSpec` defines the runtime configuration of Envoy Gateway. The name `ControlPlaneSpec` is being used
+instead of `ControlPlane` since the API may be embedded in a higher-level API that is represented as a Kubernetes
+resource.
+
+The `v1alpha1` version and `gateway.envoyproxy.io` API group get generated:
+```go
+// gateway/api/v1alpha1/doc.go
+
+// Package v1alpha1 contains API Schema definitions for the gateway.envoyproxy.io API group.
+//
+// +groupName=gateway.envoyproxy.io
+package v1alpha1
+```
+
+The initial `ControlPlaneSpec` API being proposed:
+```go
+// gateway/api/v1alpha1/controlplane.go
+
+package valpha1
+
+import (
+	"github.com/gateway/pkg/provider/file"
+	"github.com/gateway/pkg/provider/kubernetes"
+)
+
+// ControlPlaneSpec defines the desired state of Envoy Gateway configuration.
+type ControlPlaneSpec struct {
+	// Gateway defines Gateway-API specific configuration. If unset, default
+	// configuration parameters will apply.
+	//
+	// +optional
+	Gateway *Gateway `json:"gateway,omitempty"`
+
+	// Providers define provider configuration. If unspecified, the Kubernetes
+	// provider is used with default parameters.
+	//
+	// +optional
+	Providers *Providers `json:"providers,omitempty"`
+
+	// XdsServer defines the xDS Server configuration parameters. If unspecified,
+	// default configuration parameters are applied.
+	//
+	// +optional
+	XdsServer XdsServer `json:"xdsServer,omitempty"`
+}
+
+// Gateway defines desired Gateway API configuration of Envoy Gateway.
+type Gateway struct {
```

Comment:
shouldnt this be called `GatewayClass` if its referencing k8s gateway api GatewayClass fields

Reply:
@arkodg `GatewayClass` would imply a top-level field for configuring GatewayClass-related config. I suggest that all Gateway API-related configuration is grouped under the `Gateway` API type. This API type will be used to express Gateway controller, HTTPRoute controller, etc. configuration. For example:
```yaml
controlPlane:
  gateway:
    controllerName: foo
    bar: baz
    ...
```

`controlplane.gateway.className` is the name of the controller that is managing Gateways of this class. I feel like the API structure aligns well with the meaning but I'm open to suggestions.

Reply:
sg